File: patterns/support/python/contract_support.py
from __future__ import annotations

import asyncio
import copy
import inspect
import json
from pathlib import Path


def _resolve(value):
    if inspect.isawaitable(value):
        return asyncio.run(value)
    return value


def run_contract(test_file, evaluate):
    contract_path = Path(test_file).resolve().parents[1] / "fixtures" / "contract.json"
    contract = json.loads(contract_path.read_text(encoding="utf-8"))

    for case in contract["cases"]:
        try:
            result = _resolve(evaluate(copy.deepcopy(case["input"])))
        except Exception as error:
            if "expected_error" not in case:
                raise AssertionError(f"{case['id']}: unexpected error: {error}") from error
            actual_code = getattr(error, "code", None)
            assert actual_code == case["expected_error"]["code"], (
                f"{case['id']}: error code {actual_code!r} != "
                f"{case['expected_error']['code']!r}"
            )
        else:
            assert "expected_error" not in case, (
                f"{case['id']}: expected error {case['expected_error']['code']!r}, "
                f"got result {result!r}"
            )
            assert result == case["expected"], (
                f"{case['id']}: result {result!r} != {case['expected']!r}"
            )

File: patterns/support/js/contract-support.mjs
import assert from "node:assert/strict";
import { readFile } from "node:fs/promises";

export async function runContract(testModuleUrl, evaluate) {
  const contract = JSON.parse(
    await readFile(new URL("../fixtures/contract.json", testModuleUrl), "utf8"),
  );

  for (const scenario of contract.cases) {
    try {
      const result = await evaluate(structuredClone(scenario.input));
      assert.equal(
        scenario.expected_error,
        undefined,
        `${scenario.id}: expected an error, got ${JSON.stringify(result)}`,
      );
      assert.deepEqual(result, scenario.expected, scenario.id);
    } catch (error) {
      if (scenario.expected_error === undefined) {
        throw error;
      }
      assert.equal(error.code, scenario.expected_error.code, scenario.id);
    }
  }
}
